types/express-serve-static-core/index.d.ts

```diff
@@ -61,7 +61,7 @@ export interface RequestHandler<
         req: Request<P, ResBody, ReqBody, ReqQuery, LocalsObj>,
         res: Response<ResBody, LocalsObj>,
         next: NextFunction,
-    ): void;
+    ): void | Promise<void>;
```

Comment:
Could you clarify why this change was needed? I read https://github.com/DefinitelyTyped/DefinitelyTyped/issues/50871 but I still don't understand why this change is needed.

Reply:
If you're using TypeScript-ESLint's [recommended-type-checked](https://typescript-eslint.io/users/configs/#recommended-type-checked), which comes with [no-misused-promises](https://typescript-eslint.io/rules/no-misused-promises/), you get an error whenever you create an Express endpoint with an async function.

```ts
import express from "express"

const app = express()

app.post("/", async () => {
	await new Promise<void>(resolve => setTimeout(resolve, 1000))
})
```

```sh
/home/nato/Code/github.com/coderabbitai/rabbits-lab/src/express/promise.ts
  5:15  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises

✖ 1 problem (1 error, 0 warnings)
```

This is because the type definition for `RequestHandler` incorrectly asserts that no async function can ever be given to `app.post()` and that whatever is using `RequestHandler` should always assume that the return type of that handler is exactly `void` and nothing else.

That is not true, since Express itself is able to handle promise rejections from `RequestHandler`.

The type needs to be changed so that linters no longer throw an error in figuratively every usage of Express.

Reply:
Ok, I now understand the part about lint errors, thank you.

> That is not true, since Express itself is able to handle promise rejections from RequestHandler.

Is that documented somewhere?

Reply:
Yes! The code comment I placed was copied from [expressjs.com/en/guide/error-handling](https://expressjs.com/en/guide/error-handling.html#:~:text=starting%20with%20express%205%2C%20route%20handlers%20and%20middleware%20that%20return%20a%20promise%20will%20call%20next(value)%20automatically%20when%20they%20reject%20or%20throw%20an%20error.)

There's also another example usage of an async function in a `RequestHandler` in [expressjs.com/en/guide/writing-middleware](https://expressjs.com/en/guide/writing-middleware.html#:~:text=note%20how%20next()%20is%20called%20after%20await)